source/AST/ASTVisitor.cpp

```diff
@@ -1582,58 +1461,251 @@ TraverseFunctionTemplateDecl(
     // (e.g. for an abbreviated function template with no template-head)
     if(! shouldExtract(FD))
         return true;
+    auto Template = std::make_unique<TemplateInfo>();
+    parseTemplateParams(*Template, FD);
 
-    FunctionInfo I;
-    parseTemplateParams(I.Template, FD);
+    // traverse the templated declaration according to its kind
+    return TraverseDecl(FD, std::move(Template));
+}
 
-    switch(FD->getKind())
+bool
+ASTVisitor::
+Traverse(
+    ClassScopeFunctionSpecializationDecl* D)
+{
+    if(! shouldExtract(D))
+        return true;
+
+    // for class scope explicit specializations of member function templates which
+    // are members of class templates, it is impossible to know what the
+    // primary template is until the enclosing class template is instantiated.
+    // while such declarations are valid C++ (see CWG 727 and [temp.expl.spec] p3),
+    // GCC does not consider them to be valid. consequently, we do not extract the SymbolID
+    // of the primary template. in the future, we could take a best-effort approach to find
+    // the primary template, but this is only possible when none of the candidates are dependent
+    // upon a template parameter of the enclosing class template.
+    auto Template = std::make_unique<TemplateInfo>();
+    parseTemplateArgs(*Template, D);
+
+    CXXMethodDecl* MD = D->getSpecialization();
+
+    // KRYSTIAN FIXME: is the right? should this call TraverseDecl instead?
+    return Traverse(MD, std::move(Template));
+}
+
+bool
+ASTVisitor::
+Traverse(
+    TypeAliasTemplateDecl* D)
+{
+    TypeAliasDecl* AD = D->getTemplatedDecl();
+    if(! shouldExtract(AD))
+        return true;
+
+    auto Template = std::make_unique<TemplateInfo>();
+    parseTemplateParams(*Template, AD);
+
+    return Traverse(AD, std::move(Template));
+}
+
+template<typename... Args>
+auto
+ASTVisitor::
+Traverse(Args&&...)
+{
+    llvm_unreachable("no matching Traverse overload found");
+}
+
+//------------------------------------------------
+
+template<typename... Args>
+bool
+ASTVisitor::
+TraverseDecl(
+    Decl* D,
+    Args&&... args)
+{
+    Assert(D);
+    if(D->isImplicit())
+        return true;
+
+    switch(D->getKind())
     {
-    case Decl::Function:
-        buildFunction(I, FD);
+    case Decl::Namespace:
+        this->Traverse(static_cast<
+            NamespaceDecl*>(D),
+            std::forward<Args>(args)...);
+        break;
+    case Decl::CXXRecord:
+        this->Traverse(static_cast<
+            CXXRecordDecl*>(D),
+            std::forward<Args>(args)...);
         break;
     case Decl::CXXMethod:
-        buildFunction(I, static_cast<
-            CXXMethodDecl*>(FD));
+        this->Traverse(static_cast<
+            CXXMethodDecl*>(D),
+            std::forward<Args>(args)...);
```

Comment:
Unlike my other projects I'm ok with this entire statement being on one line. Up to you.